### ISSUE TITLE
Ability for plugins to detect `:nohlsearch` or `v:hlsearch`

### ISSUE THREAD

**Issue by lewis6991** [neovim/neovim#18879]
### Feature already in Vim?

No

### Feature description

I'm currently trying to implement a plugin that complements the search highlighting (https://github.com/lewis6991/satellite.nvim/issues/11). One thing that is needed is to detect when the user executes `:nohlsearch` (either interactively, via a mapping, a callback, or in some plugin code).

### Potential Solutions:

* New autocmd event `CmdExecute`: fired when any ex command is executed.
* New autocmd event `VariableSet`: fired when any Vimscript variable is modified (like `OptionSet`).
* New autocmd event `HlSearch`: fired when `v:hlsearch` is changed.

Adding `CmdExecute` seems like the best solution, but I would really appreciate any other ideas if anyone has any.

**Comment by jamessan** [comment]
Would `dictwatcheradd()` help here?

**Comment by lewis6991** [comment]
`dictwatcheradd()` doesn't appear to work for `v:`.

**Comment by justinmk** [comment]
related: https://github.com/neovim/neovim/issues/4600

> * New autocmd event `CmdExecute`: fired when any ex command is executed.

👍  Now that we have `nvim_parse_cmd()` I think we could move forward with this. A requirement is that `v:event` (?) passes the command context to the event handler.

Previous: https://github.com/neovim/neovim/pull/1218

**Comment by lewis6991** [comment]
> There would be performance concerns, 

Yes, this is very true. This is the only thing I don't like about this solution, and what leans me towards a more single-point solution like a `HlSearch` autocmd. If we can't do this efficiently, then we shouldn't do it.

We could avoid the perf hit by storing the autocmd with each command it matches when the autocmd is set, that way the command matching is only done once as opposed to each time the command runs (as #1218 does). This has a slight downside of creating loosely-coupled state and we'll need to clean up the autocmd refs for each command when the autocmd is deleted.

>  and this also makes every line of plugin code impossible to reason about

Every line of ~plugin~ vimscript code. In Lua we already have the ability to wrap functions (`vim.cmd = my_custom_vim_cmd`), so this would be the closest analogue for vimscript/ex commands.

> unless we provide controls to disable this event (is :help 'eventignore' sufficient?)

Absolutely. I think `'eventignore'` should be sufficient.

> Would an "interactive only" flag also be useful?

Maybe, though one can already use `CmdLineLeave` for the interactive cases. Maybe just add the interactive flag to `v:event`?

**Comment by matveyt** [comment]
`nohlsearch` could be rewritten as `let v:hlsearch = 0`. So for this particular problem trapping `CmdExecute` does not look like a good idea.

**Comment by justinmk** [comment]
> `nohlsearch` could be rewritten as `let v:hlsearch = 0`.

I guess we need `v:` support in `dictwatcheradd()`. Seems slightly less horrible than `VariableSet`.

**Comment by lewis6991** [comment]
I put a bit a thought into this, and I have a new proposal.

Firstly, about adding `v:` support to `dictwatcheradd()`:

* Remember that `v:event` is in `v:` which means such a feature would allow a way to listen to all autocmds. I don't know if this is a good thing or not.
* Many variables in `v:` are very context-sensitive and are locally bound in essence. I don't think giving external access (via hooks) to these variables is a good idea.
* Many variables in `v:` are constant: `v:lua`, `v:true`, `v:false`, `v:version`.
* Actually implementing this is quite non-trivial, so I don't think the ROI is high enough.

### Proposal 
So... I was thinking more about a `HlSearch` autocmd and how we can generalize it, and actually I think it makes a lot of sense to introduce a more general `Search` autocmd. Whilst allowing detection of `v:hlsearch`, it would also solve several other problems I've got with [satellite.nvim](https://github.com/lewis6991/satellite.nvim) which requires hooking onto pretty much every search related event, including:

* User hitting `n`, `N`, `&` or `*`. At the moment, the plugin creates a dummy keymap in order to refresh, and I don't think it is that robust.
* User typing in the search `/` command line. We already have `CmdlineChanged/Enter/Leave` for this, but a first class event for `Search` would be nice.

There are several search related plugins that [already exist](https://github.com/rockerBOO/awesome-neovim#search) and a `Search` autocmd would probably help improve those.

In addition, we already have a `SearchWrapped` autocmd, so this new one can replace it and we can deprecate `SearchWrapped`.

Here's what the help might look like:

```
Search          After any search event. Includes:
                - Running `:nohlsearch` or setting `v:hlsearch`
                - Making a search with |n|, |N|, |&| or |*|. If the
                  search wraps then searchwrapped key is set in v:event.
                - When a search from the |/| command is run. If 'incsearch' is
                  set then the autocmd is run for every incremental search.
```

* Bullet 1 is detected by looking at `v:hlearch`.
* Bullet 2 is detected by ???. We made need to add something to `v:event`. Maybe we can simply include the search index and number of matches instead of including a `searchwrapped` key.
* Bullet 3 is detected by `vim.v.event.cmdtype`.

**Comment by justinmk** [comment]
@lewis6991 nice analysis! makes sense to me.

**Comment by mitchellwrosen** [comment]
FYI, [vim-cool](https://github.com/romainl/vim-cool) users kind of have the ability to detect "is searching" via `g:cool_is_searching`. I know it's not the same thing as a `Search` event, but it was good enough for my use case.

**Comment by nullromo** [comment]
@lewis6991 @justinmk Ongoing discussion about a possible `Search` event here: https://github.com/vim/vim/issues/17791
